File: ruby/knight/lib/identifier.rb
# frozen_string_literal: true
require_relative 'value'

module Kn
	# An identifier within Knight
	class Identifier < Value
		# The list of known variables.
		#
		# One of the few valid reasons to use a class variable:
		# - All instances need to access it
		# - It's not meant to be directly accessible by anyone else
		# - We don't have any subclasses.
		#
		# Note that just on the really off chance that someone defines an `@@env` or something
		# in `Object`, we make this class variable name very specific.
		@@kn_identifier_env = {}

		# Parse an identifier out of the steram.
		#
		# Identifiers are defined as starting with `a-z` or `_`, and then any
		# amount of `a-z`, `_`, or digits.
		def self.parse(stream)
			stream.slice! /\A[a-z_][a-z0-9_]*/ and new($&)
		end

		# Initializes the identifier.
		def initialize(ident)
			@ident = ident
		end

		# You assign an identifier's value by calling `assign`
		def assign(value)
			@@kn_identifier_env[@ident] = value
		end

		# Executing an identifier fetches it from the list of identifiers.
		def run
			@@kn_identifier_env[@ident] or raise RunError, "unknown variable '#@ident'."
		end

		def to_i = run.to_i
		def to_s = run.to_s
		def truthy? = run.truthy?
	end
end

File: python/knight/identifier.py
from __future__ import annotations
from knight import Value, Stream, RunError
from typing import Union, Dict
import re

_ENV: Dict[str, Value] = {}

class Identifier(Value):
	"""
	Represents an identifier within Knight.

	Because all identifiers in Knight are global---and don't go out of
	scope---we have a single dict that keeps track of _all_ identifiers.
	"""

	REGEX: re.Pattern = re.compile(r'[a-z_][a-z0-9_]*')
	name: str

	@classmethod
	def parse(cls, stream: Stream) -> Union[None, Identifier]:
		"""
		Parses an Identifier out from the stream.

		This returns `None` if the stream doesn't start with a lowercase
		letter, or an underscore.
		"""
		if match := stream.matches(Identifier.REGEX):
			return cls(match)
		else:
			return None

	def __init__(self, name: str):
		""" Creates a new Identifier associated with the given `name`. """
		self.name = name

	def __repr__(self) -> str:
		""" Gets a debugging mode representation of this identifier. """
		return f"Identifier({self.name})"

	def run(self) -> Value:
		"""
		Fetches the value associated with this identifier from the list
		of known identifiers.

		If the identifier has not been assigned yet (cf `assign`), then a
		`RunError` will be raised.
		"""
		if self.name in _ENV:
			return _ENV[self.name]
		else:
			raise RunError(f"unknown identifier '{self.name}'")


	def assign(self, value: Value):
		"""
		Associated the Value `value` with this identifier.

		Any previously associated value with this identifier is discarded.
		"""
		_ENV[self.name] = value
